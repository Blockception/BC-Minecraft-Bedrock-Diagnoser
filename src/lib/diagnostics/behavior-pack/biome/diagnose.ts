import { DiagnosticsBuilder, DiagnosticSeverity } from "../../../types";
import { Errors } from "../..";

/**Checks if the biome exists in the project or in vanilla, if not then a bug is reported
 * @param id
 * @param diagnoser
 * @returns
 */
export function is_biome_defined(id: string, diagnoser: DiagnosticsBuilder, namespace_required = false): boolean {

  if (namespace_required && id.split(':').length == 1) diagnoser.add(id, "A namespace is required to reference the biome", DiagnosticSeverity.error, "behaviorpack.biome.namespace_required");
  
  //Project has biome
  const anim = diagnoser.context.getProjectData().behaviors.biomes.get(id, diagnoser.project)
  if (anim === undefined) {
    Errors.missing("behaviors", "biomes", id, diagnoser);
    return false;
  }
  return true;
}
