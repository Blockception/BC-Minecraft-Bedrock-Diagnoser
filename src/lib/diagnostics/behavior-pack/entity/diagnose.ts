import { Command, ParameterType } from "bc-minecraft-bedrock-command";
import { Types } from "bc-minecraft-bedrock-types";
import { MinecraftData } from "bc-minecraft-bedrock-vanilla-data";
import { DiagnosticsBuilder, DiagnosticSeverity } from "../../../types";
import { check_definition_value, education_enabled } from "../../definitions";
import { Errors } from "../..";
import { Defined, DefinitionItem } from "bc-minecraft-bedrock-project";

/**
 * Checks if the entities exists in the project or in vanilla, if not then a bug is reported
 * @param id The entity to check
 * @param diagnoser The diagnoser
 * @returns True if the entity exists
 */
export function behaviorpack_entityid_diagnose(id: Types.OffsetWord | string, diagnoser: DiagnosticsBuilder): boolean {
  let strId = typeof id === "string" ? id : id.text;
  let event = "";
  if (strId.includes("<")) {
    event = strId.replace(strId.split("<")[0], "").slice(1, -1);
    strId = strId.split("<")[0];
  }

  //No namespace?
  if (!strId.includes(":")) strId = "minecraft:" + strId;

  //Defined in McProject
  if (check_definition_value(diagnoser.project.definitions.entity, strId, diagnoser)) {
    return true;
  }

  const entityItem = diagnoser.context.getProjectData().behaviors.entities.get(strId, diagnoser.project);
  if (entityItem === undefined) {
    Errors.missing("behaviors", "entities", strId, diagnoser, id);
    return false;
  }
  if (DefinitionItem.is(entityItem)) {
    return true;
  }
  const entity = entityItem.item;

  //Project has entity
  if (event) {
    behaviorpack_entity_event_diagnose(event, `${strId}<${event}>`, entity.events, diagnoser);
  }

  return true;
}

/**Checks if the entities exists in the project or in vanilla, if not then a bug is reported
 * @param id
 * @param diagnoser
 * @returns
 */
export function behaviorpack_entity_spawnegg_diagnose(value: Types.OffsetWord, diagnoser: DiagnosticsBuilder): void {
  const id = value.text.replace("_spawn_egg", "");

  behaviorpack_entityid_diagnose({ offset: value.offset, text: id }, diagnoser);
}

export function behaviorpack_entity_event_diagnose(
  id: string,
  path: string,
  events: Defined | string[] | undefined,
  diagnoser: DiagnosticsBuilder
) {
  if (!events) return;

  if (Defined.is(events)) {
    if (events.defined.has(id)) return;
  } else if (events.includes(id)) {
    return;
  }

  diagnoser.add(path, `Entity has no event "${id}"`, DiagnosticSeverity.warning, "behaviorpack.entity.event.missing");
}

/**Checks if the event is defined on the correct entities
 * @param data
 * @param builder
 * @param Com
 */
export function command_entity_event_diagnose(
  data: Types.OffsetWord,
  diagnoser: DiagnosticsBuilder,
  Com: Command
): void {
  const edu = education_enabled(diagnoser);
  const matches = Com.getBestMatch(edu);

  if (matches.length !== 1) return;

  const entityid_index = matches[0].parameters.findIndex((p) => p.type === ParameterType.entity);
  let entityid: string | undefined = undefined;

  if (entityid_index >= 0) {
    entityid = Com.parameters[entityid_index]?.text;
  } else {
    //TODO selector parsing?
  }

  if (entityid) {
    //Get entity
    const entity = diagnoser.context.getProjectData().projectData.behaviorPacks.entities.get(entityid);

    //Entity found
    if (entity) {
      //Events not found
      if (!entity.events.defined.has(data.text)) {
        diagnoser.add(
          data.offset,
          `Entity: ${entityid} has no event declared: ${data.text}`,
          DiagnosticSeverity.error,
          "behaviorpack.entity.event.missing"
        );
      }
    }
  }
}
