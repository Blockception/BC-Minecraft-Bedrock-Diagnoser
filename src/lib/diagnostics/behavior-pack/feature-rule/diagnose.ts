import { check_definition_value } from "../../definitions";
import { DiagnosticsBuilder, DiagnosticSeverity } from "../../../types";

export function behaviorpack_featurerule_diagnose(id: string, diagnoser: DiagnosticsBuilder): boolean {

  if (!id) return false;

  //Defined in McProject
  if (check_definition_value(diagnoser.project.definitions.feature_rule, id, diagnoser)) return true;

  //Missing namespace?
  if (!id.includes(":")) id = "minecraft:" + id;

  //Project has
  const data = diagnoser.context.getCache();
  if (data.behaviorPacks.features_rules.has(id)) return true;

  //Vanilla has entity
  // if (MinecraftData.BehaviorPack) return true;

  //Nothing then report error
  diagnoser.add(
    id,
    `Cannot find behaviorpack feature rule definition: ${id}`,
    DiagnosticSeverity.error,
    "behaviorpack.feature_rule.missing"
  );
  return false;
}