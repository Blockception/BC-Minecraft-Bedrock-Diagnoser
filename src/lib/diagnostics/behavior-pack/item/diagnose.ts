import { behaviorpack_entityid_diagnose } from "../entity";
import { check_definition_value, education_enabled } from "../../definitions";
import { DiagnosticsBuilder, DiagnosticSeverity } from "../../../types";
import { MinecraftData } from "bc-minecraft-bedrock-vanilla-data";
import { Types } from "bc-minecraft-bedrock-types";

interface Item extends Types.OffsetWord {
  data?: number;
}

export function behaviorpack_item_diagnose(value: Item|string, diagnoser: DiagnosticsBuilder): boolean {
  let id = typeof value == 'string' ? value : value.text;
  //Defined in McProject
  if (check_definition_value(diagnoser.project.definitions.item, id, diagnoser)) return true;

  //If it is an spawn egg, treat it as an entity
  if (id.endsWith("_spawn_egg")) {
    const entity = { offset: typeof value == 'string' ? 0 : value.offset, text: id.slice(0, id.length - 10) };
    return behaviorpack_entityid_diagnose(entity, diagnoser);
  }

  if (hasAny(id, diagnoser)) {
    if (typeof value == 'string') return true;
    else return checkData(value, diagnoser);
  }

  //Missing namespace?
  if (!id.includes(":")) {
    //retry
    id = "minecraft:" + id;

    if (hasAny(id, diagnoser)) {
      value = { offset: typeof value == 'string' ? 0 : value.offset, text: id };
      return checkData(value, diagnoser);
    }
  }

  //Nothing then report error
  diagnoser.add(
    value,
    `Cannot find behaviorpack item definition: ${id}`,
    DiagnosticSeverity.error,
    "behaviorpack.item.missing"
  );
  return false;
}

function hasAny(id: string, diagnoser: DiagnosticsBuilder): boolean {
  const data = diagnoser.context.getProjectData().projectData;

  if (check_definition_value(diagnoser.project.definitions.item, id, diagnoser)) return true;
  if (check_definition_value(diagnoser.project.definitions.block, id, diagnoser)) return true;

  //Project has item Or blocks
  if (data.hasItem(id)) return true;
  if (data.hasBlock(id)) return true;

  const edu = education_enabled(diagnoser);

  //Vanilla has item
  if (MinecraftData.BehaviorPack.hasItem(id, edu)) return true;
  if (MinecraftData.BehaviorPack.hasBlock(id, edu)) return true;

  return false;
}

function checkData(value: Item, diagnoser: DiagnosticsBuilder): boolean {
  const edu = education_enabled(diagnoser);

  const item = MinecraftData.BehaviorPack.getItem(value.text, edu);
  if (item && typeof value.data === "number") {
    if (value.data <= item.max_damage) {
      diagnoser.add(
        value,
        `Item data is for ${value.text} is 0..${item.max_damage}`,
        DiagnosticSeverity.error,
        "behaviorpack.item.data"
      );
    }
  }

  return true;
}
