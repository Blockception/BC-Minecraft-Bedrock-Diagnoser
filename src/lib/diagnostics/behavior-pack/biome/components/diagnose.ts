import { ComponentBehavior } from "bc-minecraft-bedrock-types/lib/minecraft/components";
import { DiagnosticSeverity, DocumentDiagnosticsBuilder } from "../../../../types";
import { Context } from "../../../../utility/components";
import { component_error, ComponentCheck, components_check } from "../../../../utility/components/checks";
import { Internal } from "bc-minecraft-bedrock-project";
import { is_block_defined } from '../../block';
import { is_biome_defined } from '../diagnose';

/**
 *
 * @param container
 * @param context
 * @param diagnoser
 */
export function behaviorpack_diagnose_biome_components(
  container: ComponentBehavior,
  context: Context<Internal.BehaviorPack.Biome>,
  diagnoser: DocumentDiagnosticsBuilder
): void {
  components_check(container, context, diagnoser, component_test);
}

const component_test: Record<string, ComponentCheck<Internal.BehaviorPack.Biome>> = {
  "minecraft:surface_parameters": deprecated_component("minecraft:surface_builder"),
  "minecraft:frozen_ocean_surface": deprecated_component("minecraft:surface_builder"),
  "minecraft:mesa_surface": deprecated_component("minecraft:surface_builder"),
  "minecraft:swamp_surface": deprecated_component("minecraft:surface_builder"),
  "minecraft:the_end_surface": deprecated_component("minecraft:surface_builder"),
  "minecraft:climate": (name, component, context, diagnoser) => {
    const particles = ["ash", "blue_spores", "red_spores", "white_ash"]
    Object.keys(component).forEach(key => {
      if (particles.includes(key)) diagnoser.add(key, "This capability has been moved to the client_biome.json", DiagnosticSeverity.error, "behaviorpack.biome.components.climate.particles");
    })
  },
  "minecraft:overworld_generation_rules": (name, component, context, diagnoser) => {
    if (!context.source['minecraft:biome'].description.identifier.startsWith('minecraft:')) return component_error("Pre Caves and cliffs components do nothing with biome generation and will return a content error when used in custom biomes", "behaviorpack.biome.components.pre_1.17_component")
  },
  "minecraft:multinoise_generation_rules": (name, component, context, diagnoser) => {
    if (!context.source['minecraft:biome'].description.identifier.startsWith('minecraft:')) return component_error("Pre Caves and cliffs components do nothing with biome generation and will return a content error when used in custom biomes", "behaviorpack.biome.components.pre_1.17_component")
  },
  "minecraft:surface_builder": (name, component, context, diagnoser) => {
    const builder = component.builder
    const properties = ["sea_floor_material", "foundation_material", "mid_material", "top_material", "sea_material"]
    properties.forEach(key => {
      if (typeof builder[key] === 'string') is_block_defined(builder[key], diagnoser);
    })
  },
  "minecraft:replace_biomes": (name, component, context, diagnoser) => {
    if (Array.isArray(component.replacements)) component.replacements.forEach((entry: any) => {
      entry.targets.forEach((id: string) => {
        is_biome_defined(id, diagnoser, true) // TODO: Is this limited to replacing vanilla biomes only?
      });
    })
  }
};

function deprecated_component(replacement?: string) {
  const str = replacement ? ", replace with " + replacement : "";

  return component_error(
    `This component is no longer supported${str}. You are recommended to use the latest format version.`,
    "behaviorpack.biome.components.deprecated"
  );
}