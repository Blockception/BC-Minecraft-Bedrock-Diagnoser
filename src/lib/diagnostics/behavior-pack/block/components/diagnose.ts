import { ComponentBehavior } from "bc-minecraft-bedrock-types/lib/minecraft/components";
import { DiagnosticSeverity, DocumentDiagnosticsBuilder } from "../../../../types";
import { Context } from "../../../../utility/components";
import { component_error, ComponentCheck, components_check } from "../../../../utility/components/checks";
import { diagnose_block_culling_geo_and_rules } from "../../../resource-pack/block-culling";
import { model_is_defined } from "../../../resource-pack/model/diagnose";
import { behaviorpack_loot_table_diagnose } from "../../loot-table";
import { is_block_defined } from "../diagnose";
import { Internal } from "bc-minecraft-bedrock-project";
import { FormatVersion } from "bc-minecraft-bedrock-types/lib/minecraft";
import { Vanilla } from "bc-minecraft-bedrock-vanilla-data";

/**
 *
 * @param container
 * @param context
 * @param diagnoser
 */
export function behaviorpack_diagnose_block_components(
  container: ComponentBehavior,
  context: Context<Internal.BehaviorPack.Block>,
  diagnoser: DocumentDiagnosticsBuilder
): void {
  components_check(container, context, diagnoser, component_test);
}

const component_test: Record<string, ComponentCheck<Internal.BehaviorPack.Block>> = {
  // deprecated
  "minecraft:aim_collision": deprecated_component("minecraft:selection_box"),
  "minecraft:block_collision": deprecated_component("minecraft:selection_box"),
  "minecraft:block_light_absorption": deprecated_component("minecraft:light_dampening"),
  "minecraft:block_light_emission": deprecated_component("minecraft:light_emission"),
  "minecraft:block_light_filter": deprecated_component("minecraft:light_dampening"),
  "minecraft:breakonpush": deprecated_component(),
  "minecraft:breathability": deprecated_component(),
  "minecraft:creative_category": deprecated_component("description.menu_category"),
  "minecraft:destroy_time": deprecated_component("minecraft:destructible_by_mining"),
  "minecraft:entity_collision": deprecated_component("minecraft:collision_box"),
  "minecraft:explosion_resistance": deprecated_component("minecraft:destructible_by_explosion"),
  "minecraft:immovable": deprecated_component(),
  "minecraft:on_fall_on": deprecated_component("minecraft:custom_components"),
  "minecraft:on_interact": deprecated_component("minecraft:custom_components"),
  "minecraft:on_placed": deprecated_component("minecraft:custom_components"),
  "minecraft:on_player_destroyed": deprecated_component("minecraft:custom_components"),
  "minecraft:on_player_placing": deprecated_component("minecraft:custom_components"),
  "minecraft:on_step_off": deprecated_component("minecraft:custom_components"),
  "minecraft:on_step_on": deprecated_component("minecraft:custom_components"),
  "minecraft:onlypistonpush": deprecated_component(),
  "minecraft:partial_visibility": deprecated_component("minecraft:geometry.bone_visibility"),
  "minecraft:pick_collision": deprecated_component("minecraft:selection_box"),
  "minecraft:preventsjumping": deprecated_component(),
  "minecraft:queued_ticking": deprecated_component("minecraft:custom_components"),
  "minecraft:random_ticking": deprecated_component("minecraft:custom_components"),
  "minecraft:rotation": deprecated_component("minecraft:transformation"),
  "minecraft:unit_cube": deprecated_component("geometry.minecraft:full_block"),
  "minecraft:unwalkable": deprecated_component(),

  // General
  "minecraft:destructible_by_mining": (name, component, context, diagnoser) => {
    const destroyTime = component.seconds_to_destroy;
    if (!destroyTime) return;

    component.item_specific_speeds?.forEach((specific_speed: any) => {
      if (specific_speed.destroy_speed && specific_speed.destroy_speed > destroyTime) {
        diagnoser.add(
          specific_speed.destroy_speed,
          `Item specific destroy speed ${specific_speed.destroy_speed} cannot be higher than block's destroy time ${destroyTime}`,
          DiagnosticSeverity.error,
          "behaviorpack.block.components.fast_break_speed"
        );
      }
    });
  },
  "minecraft:placement_filter": (name, component, context, diagnoser) => {
    for (const condition of component.conditions) {
      condition.block_filter
        ?.map((block: string | { name: string } | {tags: string}) => (typeof block === "string" ? block : ("name" in block ? block.name : undefined)))
        .forEach((block: string | undefined) => { if (typeof block == 'string') is_block_defined(block, diagnoser) });
    }
  },
  "minecraft:geometry": (name, component, context, diagnoser) => {
    try {
      if (
        !context.components.includes("minecraft:material_instances") &&
        FormatVersion.isGreaterOrEqualThan(context.source.format_version as FormatVersion, [1, 21, 80])
      )
        diagnoser.add(
          name,
          `"minecraft:geometry" requires "minecraft:material_instances" in format versions >= 1.21.80`,
          DiagnosticSeverity.error,
          "behaviorpack.block.components.material_instances_x_geometry"
        );
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
    } catch (err) {
      // Leaving empty as the base diagnoser should flag an invalid format version
    }

    if (typeof component === "string") {
      model_is_defined(component, diagnoser);
    } else if (typeof component === "object") {
      if (component.value) model_is_defined(component.value, diagnoser);
      if (component.identifier) model_is_defined(component.identifier, diagnoser);
      if (component.culling && component.identifier)
        diagnose_block_culling_geo_and_rules(component.identifier, component.culling, diagnoser);
    }
  },
  "minecraft:loot": (name, component, context, diagnoser) => {
    if (typeof component === "string") behaviorpack_loot_table_diagnose(component, diagnoser);
  },
  "minecraft:material_instances": (name, component, context, diagnoser) => {
    try {
      if (
        !context.components.includes("minecraft:geometry") &&
        FormatVersion.isGreaterOrEqualThan(context.source.format_version as FormatVersion, [1, 21, 80])
      )
        diagnoser.add(
          name,
          `"minecraft:material_instances" requires "minecraft:geometry" in format versions >= 1.21.80`,
          DiagnosticSeverity.error,
          "behaviorpack.block.components.material_instances_x_geometry"
        );
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
    } catch (err) {
      // Leaving empty as the base diagnoser should flag an invalid format version
    }

    Object.keys(component)
      .map((value) => component[value].texture)
      .forEach((textureId) => {
        if (
          !diagnoser.context
            .getProjectData()
            .projectData.resourcePacks.terrainTextures.find((val) => val.id == textureId) &&
          !Vanilla.ResourcePack.TextureTerrain.includes(textureId)
        ) {
          diagnoser.add(
            textureId,
            `Texture reference "${textureId}" was not defined in terrain_texture.json`,
            DiagnosticSeverity.error,
            "behaviorpack.block.components.texture_not_found"
          );
        }
      });

    if (new Set(Object.keys(component).map(value => component[value].render_method)).size > 1) diagnoser.add(name,
      "Custom blocks were never intended to support multiple different render_method inside this component",
      DiagnosticSeverity.error,
      "behaviorpack.block.components.multiple_render_methods");
  },
  "minecraft:custom_components": (name, component, context, diagnoser) => {
    try {
      if (FormatVersion.isGreaterThan(context.source.format_version as FormatVersion, [1, 21, 90]))
        diagnoser.add(
          context.source.format_version,
          `'minecraft:custom_components' is deprecated in versions after 1.21.80`,
          DiagnosticSeverity.warning,
          "behaviorpack.block.components.custom_components_deprecation"
        );
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
    } catch (err) {
      // Leaving empty as the base diagnoser should flag an invalid format version
    }
  },
  "minecraft:item_visual": (name, component, context, diagnoser) => {
    minimum_version_required(context.source, name, [1, 21, 50], diagnoser);
  },
  "minecraft:liquid_detection": (name, component, context, diagnoser) => {
    minimum_version_required(context.source, name, [1, 21, 50], diagnoser);
  },
  "minecraft:redstone_conductivity": (name, component, context, diagnoser) => {
    minimum_version_required(context.source, name, [1, 21, 30], diagnoser);
  },
  "minecraft:replaceable": (name, component, context, diagnoser) => {
    minimum_version_required(context.source, name, [1, 21, 60], diagnoser);
  },
  "minecraft:movable": (name, component, context, diagnoser) => {
    minimum_version_required(context.source, name, [1, 21, 100], diagnoser);
  },
};

function minimum_version_required(
  block: Internal.BehaviorPack.Block,
  name: string,
  version: [number, number, number],
  diagnoser: DocumentDiagnosticsBuilder
) {
  try {
    if (FormatVersion.isLessThan(block.format_version as FormatVersion, version)) {
      diagnoser.add(
        name,
        `${name} requires a format version of ${version.join(".")} or greater to use.`,
        DiagnosticSeverity.error,
        "behaviorpack.block.components." + name.split(":")[1]
      );
    }
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
  } catch (err) {
    // Leaving empty as the base diagnoser should flag an invalid format version
  }
}

function deprecated_component(replacement?: string) {
  const str = replacement ? ", replace with " + replacement : "";
  return component_error(
    "This component is no longer supported" + str + ". You are recommended to use the latest format version.",
    "behaviorpack.block.components.deprecated"
  );
}
