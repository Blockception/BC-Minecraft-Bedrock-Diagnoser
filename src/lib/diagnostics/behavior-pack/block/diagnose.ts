import { Minecraft, Types } from "bc-minecraft-bedrock-types";
import { MinecraftData } from "bc-minecraft-bedrock-vanilla-data";
import { DiagnosticsBuilder, DiagnosticSeverity } from "../../../types";
import { check_definition_value, education_enabled } from "../../definitions";
import { behaviorpack_check_blockstates } from "../block-state/diagnose";
import { Errors } from "../..";

/**
 *
 * @param blockDescriptor
 * @param diagnoser
 */
export function behaviorpack_check_blockdescriptor(
  blockDescriptor: Types.OffsetWord,
  diagnoser: DiagnosticsBuilder
): void {
  behaviorpack_check_blockid_from_descriptor(blockDescriptor, diagnoser);
  behaviorpack_check_blockstates(blockDescriptor, diagnoser);
}

export function behaviorpack_check_blockid_from_descriptor(
  blockDescriptor: Types.OffsetWord,
  diagnoser: DiagnosticsBuilder
): boolean {
  return is_block_defined(Minecraft.Block.getId(blockDescriptor.text), diagnoser);
}

/**Checks if the blocks exists in the project or in vanilla, if not then a bug is reported
 * @param id
 * @param diagnoser
 * @returns
 */
export function is_block_defined(id: string, diagnoser: DiagnosticsBuilder): boolean {
  //Project has block
  const anim = diagnoser.context.getProjectData().behaviors.blocks.get(id, diagnoser.project);
  if (anim === undefined) {
    Errors.missing("behaviors", "blocks", id, diagnoser);
    return false;
  }
  return true;
}
