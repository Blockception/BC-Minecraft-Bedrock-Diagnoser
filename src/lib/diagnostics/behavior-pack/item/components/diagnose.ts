import { Internal } from "bc-minecraft-bedrock-project";
import { FormatVersion } from 'bc-minecraft-bedrock-types/lib/minecraft';
import { ComponentBehavior } from "bc-minecraft-bedrock-types/lib/minecraft/components";
import { DiagnosticSeverity, DocumentDiagnosticsBuilder } from "../../../../types";
import { Context } from "../../../../utility/components";
import { component_error, ComponentCheck, components_check } from "../../../../utility/components/checks";
import { minecraft_get_item } from "../../../minecraft/items";
import { resourcepack_particle_diagnose } from '../../../resource-pack/particle';
import { is_block_defined } from "../../block";
import { behaviorpack_entityid_diagnose } from "../../entity";
import { behaviorpack_item_diagnose } from "../diagnose";

/**
 *
 * @param container
 * @param context
 * @param diagnoser
 */
export function behaviorpack_diagnose_item_components(
  container: ComponentBehavior,
  context: Context<Internal.BehaviorPack.Item>,
  diagnoser: DocumentDiagnosticsBuilder
): void {
  components_check(container, context, diagnoser, component_test);
}

const component_test: Record<string, ComponentCheck<Internal.BehaviorPack.Item>> = {
  "minecraft:armor": deprecated_component("minecraft:wearable"),
  "minecraft:chargeable": deprecated_component("minecraft:custom_components"),
  "minecraft:creative_category": deprecated_component("description.menu_category"),
  "minecraft:dye_powder": deprecated_component(),
  "minecraft:knockback_resistance": deprecated_component(),
  "minecraft:render_offsets": deprecated_component("attatchables"),
  "minecraft:animates_in_toolbar": deprecated_component(),
  "minecraft:ignores_permission": deprecated_component(),
  "minecraft:explodable": deprecated_component(),
  "minecraft:foil": deprecated_component("minecraft:glint"),
  "minecraft:mining_speed": deprecated_component(),
  "minecraft:mirrored_art": deprecated_component(),
  "minecraft:on_use_on": deprecated_component("minecraft:custom_components"),
  "minecraft:on_use": deprecated_component("minecraft:custom_components"),
  "minecraft:weapon": deprecated_component("minecraft:custom_components"),
  "minecraft:use_duration": deprecated_component(),
  "minecraft:fertilizer": deprecated_component(),
  "minecraft:frame_count": deprecated_component(),
  "minecraft:requires_interact": deprecated_component(),
  "minecraft:map": deprecated_component(),
  "minecraft:shears": deprecated_component(),
  "minecraft:bucket": deprecated_component(),
  saddle: deprecated_component(),
  "minecraft:entity_placer": (name, component, context, diagnoser) => {
    if (Array.isArray(component.dispense_on))
      component.dispense_on.forEach((block: string | { name: string }) => {
        if (typeof block == 'object' && 'name' in block) is_block_defined(block.name, diagnoser)
        else if (typeof block == 'string') is_block_defined(block, diagnoser);
      });
    if (Array.isArray(component.use_on))
      component.use_on.forEach((block: string | { name: string }) => {
        if (typeof block == 'object' && 'name' in block) is_block_defined(block.name, diagnoser)
        else if (typeof block == 'string') is_block_defined(block, diagnoser);
      });
    if (component.entity) behaviorpack_entityid_diagnose(component.entity, diagnoser);
  },
  "minecraft:block_placer": (name, component, context, diagnoser) => {
    if (Array.isArray(component.use_on))
      component.use_on.forEach((block: string | { name: string }) => {
        if (typeof block == 'object' && 'name' in block) is_block_defined(block.name, diagnoser)
        else if (typeof block == 'string') is_block_defined(block, diagnoser);
      });
    if (component.replace_block_item && context.source['minecraft:item'].description.identifier != component.block) diagnoser.add(`minecraft:block_placer/block/${component.block}`,
      `${component.block} and ${context.source['minecraft:item'].description.identifier} need to match when trying to replace the block item`,
      DiagnosticSeverity.error,
      'behaviorpack.item.components.replace_block_ids_dont_match')
    if (component.block) {
      if (typeof component.block == 'object' && 'name' in component.block) is_block_defined((component.block as { name: string }).name, diagnoser)
      else if (typeof component.block == 'string') is_block_defined(component.block, diagnoser);
    }
  },
  "minecraft:projectile": (name, component, context, diagnoser) => {
    if (component.projectile_entity) behaviorpack_entityid_diagnose(component.projectile_entity, diagnoser);
  },
  "minecraft:repairable": (name, component, context, diagnoser) => {
    if (Array.isArray(component.repair_items))
      component.repair_items.forEach((repairEntry: any) => {
        if (Array.isArray(repairEntry.items))
          repairEntry.items.forEach((item: string) => {
            if (typeof item == 'string') behaviorpack_item_diagnose(minecraft_get_item(item, diagnoser.document), diagnoser);
          });
      });
  },
  "minecraft:icon": (name, component, context, diagnoser) => {
    let textureId: string | undefined = undefined;
    if (typeof component == 'string') textureId = component
    else if (typeof component.texture == 'string') textureId = component.texture
    if (textureId !== undefined) {
      if (!diagnoser.context.getProjectData().projectData.resourcePacks.itemTextures.find(val => val.id == textureId))
        diagnoser.add(textureId,
          `Texture reference "${textureId}" was not defined in item_texture.json`,
          DiagnosticSeverity.error,
          'behaviorpack.item.components.texture_not_found')
    } else {
      Object.keys(component.textures)?.forEach(value => {
        const textureId = component.textures[value];
        if (!diagnoser.context.getProjectData().projectData.resourcePacks.itemTextures.find(val => val.id == textureId))
          diagnoser.add(textureId,
            `Texture reference "${textureId}" was not defined in item_texture.json`,
            DiagnosticSeverity.error,
            'behaviorpack.item.components.texture_not_found')
      })
    }

  },
  "minecraft:custom_components": (name, component, context, diagnoser) => {
    try {
      if (FormatVersion.isLessThan(FormatVersion.parse(context.source.format_version), [1,21,10])) diagnoser.add(context.source.format_version,
        `To use custom components, a minimum format version of 1.21.10 is required`,
        DiagnosticSeverity.error,
        'behaviorpack.item.components.custom_components_min_version')
      
      if (FormatVersion.isGreaterThan(FormatVersion.parse(context.source.format_version), [1,21,90])) diagnoser.add(context.source.format_version,
        `'minecraft:custom_components' is deprecated in versions after 1.21.80`,
        DiagnosticSeverity.warning,
        'behaviorpack.item.components.custom_components_deprecation')
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
    } catch (err) {
      // Leaving empty as the base diagnoser should flag an invalid format version
    }
  },
  "minecraft:durability_sensor": (name, component, context, diagnoser) => {
    if (!component.particle_type || !(typeof component.particle_type == 'string')) return;
    resourcepack_particle_diagnose(component.particle_type, diagnoser)
  },
  "minecraft:rarity": (name, component, context, diagnoser) => {
    try {
      if (FormatVersion.isLessThan(FormatVersion.parse(context.source.format_version), [1,21,30])) diagnoser.add(context.source.format_version,
        `To use "minecraft:rarity", a minimum format version of 1.21.30 is required`,
        DiagnosticSeverity.warning,
        'behaviorpack.item.components.rarity')
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
    } catch (err) {
      // Leaving empty as the base diagnoser should flag an invalid format version
    }
  },
};

function deprecated_component(replacement?: string) {
  const str = replacement ? ", replace with " + replacement : "";
  return component_error(
    "This component is no longer supported" + str + ". You are recommended to use the latest format version.",
    "behaviorpack.item.components.deprecated"
  );
}
