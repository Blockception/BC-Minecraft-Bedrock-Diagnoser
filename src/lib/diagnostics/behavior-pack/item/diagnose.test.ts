import { ItemDefinition } from "./diagnose";

const validIds = ["minecraft:stick", "stick", "stick:0", "namespace:item:variant", "namespace:item"];

describe("ItemDefinition", () => {
  test.each(validIds)("can be parsed correctly: %s", (id) => {
    expect(ItemDefinition.parse(id)).toMatchSnapshot();
  });
});
