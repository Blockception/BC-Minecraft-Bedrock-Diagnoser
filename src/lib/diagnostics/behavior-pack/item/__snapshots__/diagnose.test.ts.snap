// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`ItemDefinition can be parsed correctly: minecraft:stick 1`] = `
{
  "id": "stick",
  "namespace": "minecraft",
  "variant": "",
}
`;

exports[`ItemDefinition can be parsed correctly: namespace:item 1`] = `
{
  "id": "item",
  "namespace": "namespace",
  "variant": "",
}
`;

exports[`ItemDefinition can be parsed correctly: namespace:item:variant 1`] = `
{
  "id": "item",
  "namespace": "namespace",
  "variant": "variant",
}
`;

exports[`ItemDefinition can be parsed correctly: stick 1`] = `
{
  "id": "stick",
  "namespace": "minecraft",
  "variant": "",
}
`;

exports[`ItemDefinition can be parsed correctly: stick:0 1`] = `
{
  "id": "stick",
  "namespace": "minecraft",
  "variant": "0",
}
`;
